File: build_static.py
import os
import base64
import re
import mimetypes
from pathlib import Path

def read_file(filepath):
    with open(filepath, 'rb') as f:
        return f.read()

def to_data_url(content, mimetype):
    b64 = base64.b64encode(content).decode('utf-8')
    return f"data:{mimetype};base64,{b64}"

def main():
    # Get the build directory
    dist_dir = Path('dist/static')

    # Read the main HTML file
    with open(dist_dir / 'index.html', 'r', encoding='utf-8') as f:
        html = f.read()

    # Find all assets in the HTML
    css_links = re.findall(r'<link[^>]*href="([^"]*\.css)"[^>]*>', html)
    js_scripts = re.findall(r'<script[^>]*src="([^"]*\.js)"[^>]*>', html)

    print("Found assets:")
    print("CSS:", css_links)
    print("JS:", js_scripts)

    # Convert CSS files to data URLs
    for css_link in css_links:
        try:
            css_path = dist_dir / css_link.lstrip('./')
            css_content = read_file(css_path)
            data_url = to_data_url(css_content, 'text/css')

            # Replace the link tag with an embedded style tag
            html = html.replace(
                f'<link rel="stylesheet" crossorigin href="{css_link}">',
                f'<style type="text/css">{css_content.decode("utf-8")}</style>'
            )
            print(f"Embedded CSS file: {css_path}")
        except Exception as e:
            print(f"Error processing CSS {css_link}: {e}")

    # Convert JS files to data URLs
    for js_src in js_scripts:
        try:
            js_path = dist_dir / js_src.lstrip('./')
            js_content = read_file(js_path)
            data_url = to_data_url(js_content, 'application/javascript')

            # Replace the script src with the data URL
            html = html.replace(
                f'src="{js_src}"',
                f'src="{data_url}"'
            )
            print(f"Embedded JS file: {js_path}")
        except Exception as e:
            print(f"Error processing JS {js_src}: {e}")

    # Remove any remaining crossorigin attributes
    html = html.replace(' crossorigin', '')

    # Write the self-contained file
    output_path = dist_dir / 'standalone.html'
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(html)

    print(f"\nCreated self-contained HTML file at: {output_path}")
    print("You can now open this file directly in your browser!")

if __name__ == '__main__':
    main()

File: build-static.mjs
import fs from 'fs/promises';
import path from 'path';

async function main() {
  try {
    // Use the correct build directory
    const distDir = path.join(process.cwd(), 'dist/public');
    console.log(`Working directory: ${process.cwd()}`);
    console.log(`Build directory: ${distDir}`);

    // Read index.html
    console.log('Reading index.html...');
    const indexPath = path.join(distDir, 'index.html');
    let html = await fs.readFile(indexPath, 'utf-8');

    // Find all CSS and JS files referenced in the HTML
    const cssFiles = (html.match(/href="[^"]*\.css"/g) || [])
      .map(match => match.match(/href="([^"]*)"/)[1]);
    const jsFiles = (html.match(/src="[^"]*\.js"/g) || [])
      .map(match => match.match(/src="([^"]*)"/)[1]);

    console.log('Found assets:', { cssFiles, jsFiles });

    // Inline CSS files
    for (const cssFile of cssFiles) {
      const cssPath = path.join(distDir, cssFile.replace(/^\.\//, ''));
      console.log(`Reading CSS: ${cssPath}`);
      const cssContent = await fs.readFile(cssPath, 'utf-8');

      // Replace the link tag with an inline style tag
      html = html.replace(
        `<link rel="stylesheet" crossorigin href="${cssFile}">`,
        `<style type="text/css">${cssContent}</style>`
      );
      console.log(`Inlined CSS: ${cssPath}`);
    }

    // Inline JavaScript files
    for (const jsFile of jsFiles) {
      const jsPath = path.join(distDir, jsFile.replace(/^\.\//, ''));
      console.log(`Reading JS: ${jsPath}`);
      const jsContent = await fs.readFile(jsPath, 'utf-8');

      // Replace the script tag with an inline script tag
      html = html.replace(
        `<script type="module" crossorigin src="${jsFile}"></script>`,
        `<script type="module">${jsContent}</script>`
      );
      console.log(`Inlined JS: ${jsPath}`);
    }

    // Remove any remaining crossorigin attributes
    html = html.replace(/ crossorigin/g, '');

    // Write the self-contained file
    const outputPath = path.join(distDir, 'standalone.html');
    await fs.writeFile(outputPath, html, 'utf-8');

    const stats = await fs.stat(outputPath);
    console.log(`Created standalone.html (${stats.size} bytes)`);
    console.log('You can now open this file directly in your browser');

  } catch (err) {
    console.error('Build failed:', err);
    process.exit(1);
  }
}

main();